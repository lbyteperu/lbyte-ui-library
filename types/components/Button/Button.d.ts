import React, { ReactNode } from "react";
export type ButtonStyleColor = "primary" | "secondary" | "success" | "danger" | "warning" | "info" | "light" | "dark" | "link";
export type ButtonType = "button" | "submit" | "reset";
export type ButtonSize = "small" | "medium" | "large";
export type ButtonProps = {
    id?: string;
    anchor?: boolean;
    autoFocus?: boolean;
    children?: ReactNode;
    circled?: boolean;
    dataFsExclude?: boolean;
    disabled?: boolean;
    quiet?: boolean;
    filled?: boolean;
    inverted?: boolean;
    type?: ButtonType;
    "aria-haspopup"?: boolean;
    styleColor?: ButtonStyleColor;
    size?: ButtonSize;
    wrap?: boolean;
};
/**
 * Default button
 */
declare const Button: ({ children, styleColor, size, wrap }: ButtonProps) => React.JSX.Element;
export default Button;
