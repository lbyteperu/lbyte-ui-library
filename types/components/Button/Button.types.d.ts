import { ReactNode } from "react";
export interface ButtonProps {
    children?: ReactNode;
    value?: string;
    outline?: boolean;
    color: "primary" | "secondary" | "success" | "danger" | "warning" | "info" | "light" | "dark" | "link";
    size?: "small" | "medium" | "large";
    wrap?: boolean;
    quiet?: boolean;
}
