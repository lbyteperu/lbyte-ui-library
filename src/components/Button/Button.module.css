.base {
  font-family: var(--font-family-arial-sans-serif);
  border: 0;
  text-transform: uppercase;
  border-radius: 3rem;
  cursor: pointer;
  display: inline-block;
  transition: all 0.5s;
  letter-spacing: 0.0625rem;
  line-height: 1rem;
  font-weight: bold;
  padding: 0.625rem 1.25rem 0.5rem;
  margin: 0.5rem;
  color: var(--white);
  &:focus {
    transition: all 0.5s;
  }
  &:hover {
    transition: all 0.5s;
  }
}
.primary {
  background-color: var(--primary);
  &:focus {
    background-color: var(--primary-dark);
    box-shadow: 0rem 0rem 0rem 0.25rem var(--primary-shadow-light);
  }
  &:focus-within {
    background-color: var(--primary-dark);
  }
  &:visited {
    background-color: var(--primary-shadow);
  }
  &:hover {
    background-color: var(--primary-dark);
    box-shadow: 0rem 0rem 0rem 0.25rem var(--primary-shadow);
  }
  &:focus-visible {
    outline: 0rem 0rem 0rem 0.25rem var(--primary-shadow-light);
  }
}
.secondary {
  background-color: var(--secondary);
  &:focus {
    background-color: var(--secondary-dark);
    box-shadow: 0rem 0rem 0rem 0.25rem var(--secondary-shadow-light);
  }
  &:focus-within {
    background-color: var(--secondary-dark);
  }
  &:visited {
    background-color: var(--secondary-shadow);
  }
  &:hover {
    background-color: var(--secondary-dark);
    box-shadow: 0rem 0rem 0rem 0.25rem var(--secondary-shadow);
  }
  &:focus-visible {
    outline: 0rem 0rem 0rem 0.25rem var(--secondary-shadow-light);
  }
}
.success {
  background-color: var(--success);
  &:focus {
    background-color: var(--success-dark);
    box-shadow: 0rem 0rem 0rem 0.25rem var(--success-shadow-light);
  }
  &:focus-within {
    background-color: var(--success-dark);
  }
  &:visited {
    background-color: var(--success-shadow);
  }
  &:hover {
    background-color: var(--success-dark);
    box-shadow: 0rem 0rem 0rem 0.25rem var(--success-shadow);
  }
  &:focus-visible {
    outline: 0rem 0rem 0rem 0.25rem var(--success-shadow-light);
  }
}
.danger {
  background-color: var(--danger);
  &:focus {
    background-color: var(--danger-dark);
    box-shadow: 0rem 0rem 0rem 0.25rem var(--danger-shadow-light);
  }
  &:focus-within {
    background-color: var(--danger-dark);
  }
  &:visited {
    background-color: var(--danger-shadow);
  }
  &:hover {
    background-color: var(--danger-dark);
    box-shadow: 0rem 0rem 0rem 0.25rem var(--danger-shadow);
  }
  &:focus-visible {
    outline: 0rem 0rem 0rem 0.25rem var(--danger-shadow-light);
  }
}
.warning {
  color: var(--black);
  background-color: var(--warning);
  &:focus {
    background-color: var(--warning-dark);
    box-shadow: 0rem 0rem 0rem 0.25rem var(--warning-shadow-light);
  }
  &:focus-within {
    background-color: var(--warning-dark);
  }
  &:visited {
    background-color: var(--warning-shadow);
  }
  &:hover {
    background-color: var(--warning-dark);
    box-shadow: 0rem 0rem 0rem 0.25rem var(--warning-shadow);
  }
  &:focus-visible {
    outline: 0rem 0rem 0rem 0.25rem var(--warning-shadow-light);
  }
}
.info {
  color: var(--black);
  background-color: var(--info);
  &:focus {
    background-color: var(--info-dark);
    box-shadow: 0rem 0rem 0rem 0.25rem var(--info-shadow-light);
  }
  &:focus-within {
    background-color: var(--info-dark);
  }
  &:visited {
    background-color: var(--info-shadow);
  }
  &:hover {
    background-color: var(--info-dark);
    box-shadow: 0rem 0rem 0rem 0.25rem var(--info-shadow);
  }
  &:focus-visible {
    outline: 0rem 0rem 0rem 0.25rem var(--info-shadow-light);
  }
}
.light {
  color: var(--black);
  background-color: var(--light);
  &:focus {
    background-color: var(--light-dark);
    box-shadow: 0rem 0rem 0rem 0.25rem var(--light-shadow-light);
  }
  &:focus-within {
    background-color: var(--light-dark);
  }
  &:visited {
    background-color: var(--light-shadow);
  }
  &:hover {
    background-color: var(--light-dark);
    box-shadow: 0rem 0rem 0rem 0.25rem var(--light-shadow);
  }
  &:focus-visible {
    outline: 0rem 0rem 0rem 0.25rem var(--light-shadow-light);
  }
}
.dark {
  background-color: var(--black);
  &:focus {
    background-color: var(--dark-dark);
    box-shadow: 0rem 0rem 0rem 0.25rem var(--dark-shadow-light);
  }
  &:focus-within {
    background-color: var(--dark-dark);
  }
  &:visited {
    background-color: var(--dark-shadow);
  }
  &:hover {
    background-color: var(--dark-dark);
    box-shadow: 0rem 0rem 0rem 0.25rem var(--dark-shadow);
  }
  &:focus-visible {
    outline: 0rem 0rem 0rem 0.25rem var(--dark-shadow-light);
  }
}
.link {
  font-family: var(--font-family-arial-sans-serif);
  border: 1px solid transparent;
  text-decoration: underline;
  border-radius: 3rem;
  cursor: pointer;
  display: inline-block;
  transition: color 0.15s ease-in-out, background-color 0.15s ease-in-out,
    border-color 0.15s ease-in-out, box-shadow 0.15s ease-in-out;
  letter-spacing: 0.0625rem;
  line-height: 1rem;
  font-weight: bold;
  background-color: transparent;
  user-select: none;
  padding: 0.625rem 1.25rem 0.5rem;
  margin: 0.5rem;
  color: var(--link-color);
  &:focus {
    box-shadow: 0 0 0 0.25rem var(--link-shadow);
  }
  &:hover {
    color: var(--link-color-hover);
  }
  &:focus-visible {
    outline: 0 0 0 0.25rem var(--link-shadow);
  }
}
.storybook-button--small {
  padding: 0.5rem 1.5rem;
  font-size: 0.55rem;
  &:focus {
    padding: 0.5rem 1.5rem;
  }
  &:hover {
    padding: 0.5rem 1.5rem;
  }
}
.storybook-button--medium {
  padding: 0.8rem 1.8rem;
  font-size: 0.75rem;
  &:focus {
    padding: 0.8rem 1.8rem;
  }
  &:hover {
    padding: 0.8rem 1.8rem;
  }
}
.storybook-button--large {
  padding: 1rem 2rem;
  font-size: 1.1rem;
  &:focus {
    padding: 1rem 2rem;
  }
  &:hover {
    padding: 1rem 2rem;
  }
}

.wrap {
  max-width: 14rem;
}

.no-wrap {
  max-width: auto;
}

.outline {
  background-color: transparent;
  border: 1px;
  border-style: solid;
  &:hover {
    color: var(--white);
  }
  &:focus {
    background-color: transparent;
  }
  &:focus-within {
    background-color: transparent;
  }
}

.outline-primary {
  border-color: var(--primary);
  color: var(--primary);
  &:focus-within {
    color: var(--primary);
  }
}
.outline-secondary {
  border-color: var(--secondary);
  color: var(--secondary);
  &:focus-within {
    color: var(--secondary);
  }
}
.outline-success {
  border-color: var(--success);
  color: var(--success);
  &:focus-within {
    color: var(--success);
  }
}
.outline-danger {
  border-color: var(--danger);
  color: var(--danger);
  &:focus-within {
    color: var(--danger);
  }
}
.outline-warning {
  border-color: var(--warning);
  color: var(--warning);
  &:focus-within {
    color: var(--warning);
  }
}
.outline-info {
  border-color: var(--info);
  color: var(--info);
  &:focus-within {
    color: var(--info);
  }
}
.outline-light {
  border-color: var(--light);
  color: var(--light);
  &:focus-within {
    color: var(--light);
  }
  &:hover {
    color: var(--black);
  }
}
.outline-dark {
  border-color: var(--dark);
  color: var(--dark);
  &:focus-within {
    color: var(--dark);
  }
  &:hover {
    color: var(--white);
  }
}
.outline-link {
  border-color: var(--link-color);
  color: var(--link-color);
  &:focus-within {
    color: var(--link-color);
  }
  &:hover {
    color: var(--white);
    background-color: var(--link-color);
    box-shadow: 0 0 0 0.25rem var(--link-shadow);
  }
}
