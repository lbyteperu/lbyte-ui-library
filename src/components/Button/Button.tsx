import classList from "classnames";
import React, { ReactNode } from "react";
import styles from "./Button.module.css";

export type ButtonStyleColor =
  | "primary"
  | "secondary"
  | "success"
  | "danger"
  | "warning"
  | "info"
  | "light"
  | "dark"
  | "link";

export type ButtonTag = "link" | "button" | "input" | "submit" | "reset";
export type ButtonSize = "small" | "medium" | "large";

export type ButtonProps = {
  id?: string;
  anchor?: boolean;
  autoFocus?: boolean;
  children?: ReactNode;
  value?: string;
  circled?: boolean;
  dataFsExclude?: boolean;
  disabled?: boolean;
  quiet?: boolean;
  filled?: boolean;
  inverted?: boolean;
  buttonTag?: ButtonTag;
  "aria-haspopup"?: boolean;
  styleColor?: ButtonStyleColor;
  size?: ButtonSize;
  wrap?: boolean;
  href?: string;
};

/**
 * Default button
 */
const Button = ({
  children,
  styleColor,
  size,
  wrap,
  buttonTag,
  value,
  href,
}: ButtonProps) => {
  const buttonClassName = classList(
    styleColor === "link" ? "" : styles.base,
    styles[`${styleColor}`],
    styles[`storybook-button--${size}`],
    wrap ? styles["wrap"] : styles["no-wrap"]
  );

  return buttonTag === "link" ? (
    <a href={href} role="button" className={buttonClassName}>
      {children}
    </a>
  ) : buttonTag === "button" ? (
    <button className={buttonClassName} role="button">
      {children}
    </button>
  ) : buttonTag === "input" ? (
    <input
      type="button"
      role="button"
      className={buttonClassName}
      value={value}
    />
  ) : buttonTag === "submit" ? (
    <input
      type="submit"
      value={value}
      role="button"
      className={buttonClassName}
    />
  ) : buttonTag === "reset" ? (
    <input
      type="reset"
      value={value}
      role="button"
      className={buttonClassName}
    />
  ) : null;
};

export default Button;
