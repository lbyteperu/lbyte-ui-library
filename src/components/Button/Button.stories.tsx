import type { Meta, StoryObj } from "@storybook/react";
import { withRouter } from "storybook-addon-react-router-v6";

import Button from "./Button";
import React from "react";

type Story = StoryObj<typeof Button>;

/**
 * Button component
 */
export default {
  component: Button,
  decorators: [withRouter],
  tags: ["autodocs"],
  title: "Button",
} satisfies Meta<typeof Button>;

export const StyledButton: Story = {
  render: ({ ...args }) => {
    return (
      <div>
        <Button styleColor="primary" size="medium" children="Primary" />
        <Button styleColor="secondary" size="medium" children="Secondary" />
        <Button styleColor="success" size="medium" children="Success" />
        <Button styleColor="danger" size="medium" children="Danger" />
        <Button styleColor="warning" size="medium" children="Warning" />
        <Button styleColor="info" size="medium" children="Info" />
        <Button styleColor="light" size="medium" children="Light" />
        <Button styleColor="dark" size="medium" children="Dark" />
        <Button styleColor="link" size="medium" children="Link" />
      </div>
    );
  },
};

export const ButtonSizes: Story = {
  render: ({ ...args }) => {
    return (
      <div>
        <Button styleColor="primary" size="small" children="Small" />
        <Button styleColor="secondary" size="medium" children="Medium" />
        <Button styleColor="success" size="large" children="Large" />
      </div>
    );
  },
};

export const LongTextNoWrap: Story = {
  args: {
    children: "really long text not to be wrapped",
    styleColor: "secondary",
    size: "medium",
  },
};

export const LongTextWrapped: Story = {
  args: {
    children: "really long text not to be wrapped",
    styleColor: "secondary",
    size: "medium",
    wrap: true,
  },
};

export const LinkButton: Story = {
  args: {
    children: "Link",
    styleColor: "link",
    size: "medium",
  },
};

export const Medium: Story = {
  args: {
    children: "button",
    styleColor: "secondary",
    size: "medium",
  },
};

export const Small: Story = {
  args: {
    children: "button",
    styleColor: "secondary",
    size: "small",
  },
};
