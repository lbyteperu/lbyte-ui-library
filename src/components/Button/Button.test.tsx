import React from "react";
import { render, screen } from "@testing-library/react";

import Button from "./Button";

describe("Button", () => {
  test("renders the Button component", () => {
    render(<Button color="primary" />);
    expect(screen.getByRole("button")).toBeTruthy();
  });
  test("renders with classname primary", () => {
    const { container } = render(<Button color="primary" />);
    expect(container.firstChild).toMatchSnapshot(
      `<button class="base primary" role="button" />`,
    );
  });
  test("renders with size", () => {
    const { container } = render(<Button size="small" color="secondary" />);
    expect(container.firstChild).toMatchSnapshot(
      `<button class="base secondary small" role="button" />`,
    );
  });
  test("renders with wrap", () => {
    const { container } = render(<Button color="secondary" wrap />);
    expect(container.firstChild).toMatchSnapshot(
      `<button class="base secondary wrap" role="button" />`,
    );
  });
  test("renders with outline", () => {
    const { container } = render(<Button color="secondary" outline />);
    expect(container.firstChild).toMatchSnapshot(
      `<button class="base secondary outline-secondary outline" role="button" />`,
    );
  });
  test("renders with quiet", () => {
    const { container } = render(<Button color="secondary" quiet />);
    expect(container.firstChild).toMatchSnapshot(
      `<button class="base secondary quiet" role="button" />`,
    );
  });
});
