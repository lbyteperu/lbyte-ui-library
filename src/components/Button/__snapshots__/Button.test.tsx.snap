// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button renders with classname primary: <button class="base primary" role="button" /> 1`] = `
<button
  class="base primary"
  role="button"
/>
`;

exports[`Button renders with outline: <button class="base secondary outline-secondary outline" role="button" /> 1`] = `
<button
  class="base secondary outline-secondary outline"
  role="button"
/>
`;

exports[`Button renders with quiet: <button class="base secondary quiet" role="button" /> 1`] = `
<button
  class="base secondary quiet"
  role="button"
/>
`;

exports[`Button renders with size: <button class="base secondary small" role="button" /> 1`] = `
<button
  class="base secondary small"
  role="button"
/>
`;

exports[`Button renders with wrap: <button class="base secondary wrap" role="button" /> 1`] = `
<button
  class="base secondary wrap"
  role="button"
/>
`;
